Work in progress: Improve slash command core api

Shadorc opened the pull request:
**This pull request is following https://github.com/Discord4J/Discord4J/pull/854 which was abandoned due to too many merge conflicts**

**Description:**
- Implements Application Command objects 
  - [ApplicationCommand](https://github.com/Discord4J/Discord4J/pull/855/files#diff-cbfc98fa16460d6c52b6ed495a73b5381ef91214018a0ee705f3e8c5e9048ae7R1)
  - [ApplicationCommandOption](https://github.com/Discord4J/Discord4J/pull/855/files#diff-bd1402242832cf98280b41279e72f252aafe156853777750a2a449dd3c7cb181R1)
  - [ApplicationCommandOptionChoice](https://github.com/Discord4J/Discord4J/pull/855/files#diff-d16c9f6857cd23f71d6bee238673390ff2dd2f849063d0f21674bfc002b375deR1)
  - [ApplicationCommandInteraction](https://github.com/Discord4J/Discord4J/pull/855/files#diff-1a31ab00146b82bd949cda269b8942a91b91d62a37fba035289dc8b0765facffR1)
  - [ApplicationCommandInteractionOption](https://github.com/Discord4J/Discord4J/pull/855/files#diff-c2cc07f53d622758accb38872540104c43a4719721304272a77ef9794d6d5440R1)
  - [ApplicationCommandInteractionOptionValue](https://github.com/Discord4J/Discord4J/pull/855/files#diff-1807e9d40980675dfcb1d5aa70a6c063a6c6b01717f4f76f98ffe16c84d07091R1)
  - [Interaction](https://github.com/Discord4J/Discord4J/pull/855/files#diff-cfb9f1ef0ae4963db763b2bc1dd5c486df8ac3a95b3c3f14a991c899073b9db5R1)
  - [MessageInteraction](https://github.com/Discord4J/Discord4J/pull/855/files#diff-5c5c92577cc4b87177253667da9699a98cb7863b271dcc305089bb5cc9d7cab9R1)
    - [Message#getInteraction](https://github.com/Discord4J/Discord4J/pull/855/files#diff-0e9ff5f4bff304dbddc54ec065b8f0deedeef24c63674c9a5d0ed01b6ddca3a3R462)
  - [InteractionApplicationCommandCallbackSpec](https://github.com/Discord4J/Discord4J/pull/855/commits/b03546e4ccfd1c5cb557b325da4d098387c344ff)

- Implements Applications Command events (Create, Update, Delete)
  - [Updates ReactiveEventAdapter](https://github.com/Discord4J/Discord4J/pull/855/files#diff-5caf76dbc6cbc4670f2ed706296b94ceff73f3399e50c72b0af7cf4450dbe3efR162)
  - [Creates dispatch handlers](https://github.com/Discord4J/Discord4J/pull/855/files#diff-f77e106a22fbdbff188bc74a869889f6ee09c6c5a12fc846a4a01c3a99be6049R1)
  - [Adds handlers](https://github.com/Discord4J/Discord4J/pull/855/files#diff-f8c537f56b25201b13d8a6feeb23ec36289fca72d964db28e1f5133509dcfdfeR93)
  - [ApplicationCommandCreateEvent](https://github.com/Discord4J/Discord4J/pull/855/files#diff-73a5fe9fe9c78b56f0f63e5d3fac7dc8a8b35ac92667917e677bd00cadf3f5acR1)
  - [ApplicationCommandDeleteEvent](https://github.com/Discord4J/Discord4J/pull/855/files#diff-f952fce8216e7e052bd577e5c4b3bda8fdfca577fa252a055160a3be4eb2fe18R1)
  - [ApplicationCommandUpdateEvent](https://github.com/Discord4J/Discord4J/pull/855/files#diff-fcaaabc6d5914c6c7e923255659e5be491b24568bcec62743e63ca08d9aaf48dR1)
- Updates classes
  - [InteractionCreateEvent](https://github.com/Discord4J/Discord4J/pull/855/files#diff-de726ca42753b2ae1337ec68f3dbc259b3f92a1f23ab9a160d3c28c87406c41aR19)
  - [ExampleInteractionCreateEvent](https://github.com/Discord4J/Discord4J/pull/855/files#diff-350daeca4b6fe2b1a4ffd6c4513f5186017ae70f097b8a26f5e7236d78fabd69R19)
- Code clean-up
  - Replaces `Constructs a {@code Foo} with an associated ServiceMediator and Discord data.` with `Constructs a {@code Foo} with an associated {@link GatewayDiscordClient} and Discord data.`
  - [Removes unused log in InteractionCreateEvent class](https://github.com/Discord4J/Discord4J/pull/854/files#diff-de726ca42753b2ae1337ec68f3dbc259b3f92a1f23ab9a160d3c28c87406c41aL34)
  - Removes unused imports
  - Adds missing `final`

**Justification:** https://discord.com/developers/docs/interactions/slash-commands

Shadorc commented:
I have changed `InteractionCreateEvent#getUser` so it extracts the user data when the member is provided
```
public User getUser() {
    UserData userData = data.member().isAbsent() ? data.user().get() : data.member().get().user();
    return new User(getClient(), userData);
}
```

Shadorc commented:
I also need some feedback on the `ApplicationCommandInteractionOptionValue` class
https://github.com/Discord4J/Discord4J/pull/855/commits/38559c0f1dba5bda66fbc5e74745b90283389770

Shadorc commented:
I have implemented what was missing, this pull request is now ready for review.
I have deleted my previous comments to include them in the pull request description. 
It is still missing specs for followup messages but it will need some more work and I don't want to create an impossible-to-review pull request

quanticc commented on core/src/main/java/discord4j/core/object/audit/ChangeKey.java:
```diff
@@ -16,14 +16,14 @@
-import discord4j.common.util.Snowflake;
 
-import java.awt.*;
+import java.awt.Color;
```
Hmm.. this is unrelated to this PR but just noticed we still depend on java.awt.Color here

Shadorc replied:
Fixed in https://github.com/Discord4J/Discord4J/pull/862

quanticc commented on core/src/main/java/discord4j/core/object/command/ApplicationCommandInteractionOptionValue.java:
```diff
@@ -0,0 +1,93 @@
+
+    public Mono<User> asUser() {
+        if (type != ApplicationCommandOption.Type.USER.getValue()) {
+            throw new IllegalArgumentException("Option value cannot be converted as user");
```
Can you return Mono.error instead? Let's not depend on assembly behavior for these Mono returning methods

quanticc commented on core/src/main/java/discord4j/core/object/command/ApplicationCommandInteractionOptionValue.java:
```diff
@@ -0,0 +1,93 @@
+
+    public Mono<Role> asRole() {
+        if (type != ApplicationCommandOption.Type.ROLE.getValue()) {
+            throw new IllegalArgumentException("Option value cannot be converted as role");
```
Same comment about Mono.error, but also needs a check whether this value has a null guild ID

Shadorc replied:
In case `guildId` is null, should it returns the same error or something else? It seems like something would have gone wrong if guildId was null with a role type

quanticc replied:
Got it

quanticc commented on core/src/main/java/discord4j/core/object/command/ApplicationCommandInteractionOptionValue.java:
```diff
@@ -0,0 +1,93 @@
+
+    public Mono<Channel> asChannel() {
+        if (type != ApplicationCommandOption.Type.CHANNEL.getValue()) {
+            throw new IllegalArgumentException("Option value cannot be converted as channel");
```
Same comment about Mono.error

quanticc commented on core/src/main/java/discord4j/core/object/command/ApplicationCommandOptionChoice.java:
```diff
@@ -0,0 +1,77 @@
+        try {
+            return Long.parseLong(data.value());
+        } catch (NumberFormatException err) {
+            // TODO
```
Do we need to decide on something here?

Shadorc replied:
I have left this todo in case the exception class or the message was not accurate, if it's ok for you then I will remove the todo.

quanticc replied:
Looks good

quanticc commented on core/src/main/java/discord4j/core/object/command/Interaction.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @return The invoking user.
+     */
+    public User getUser() {
+        UserData userData = data.member().isAbsent() ? data.user().get() : data.member().get().user();
+        return new User(getClient(), userData);
+    }
```
This is fine. Just add a note in docs that we'll get the data from member for guild interactions.

quanticc reviewed (changes requested):
Big thanks for taking this. I think it only needs docs in some specific areas to clarify.

One nitpick, some files have 1 line change due to import reordering, are you using our meta repo style file?

Shadorc commented:
Concerning imports, I use the meta repo style file but I've been running "Optimize imports" option from IntelliJ because I had a lot of unused imports with all my tests, this is why some imports got re-ordered. I can cancel those changes if needed

darichey commented:
One small thing: I don't think we should switch out the switches in our enum `of()` methods for the array filter. There are known performance issues with that method. It would be better to benchmark and decide if we can afford it in a different PR.

quanticc reviewed (changes requested):
After discussion with @darichey we've decided to update our stance on this PR, considering we have to eventually merge this forward to master branch that has wide store changes, we want to keep its surface as small as possible. Two requests then:

Can you remove all enum `of` changes? Partly from a performance warning (Channel.of is called by d4j on payload handling) but mostly because it's unrelated to slash command API.

Avoid calling "Reformat code" on the project, and only apply it to files that have actual changes.

Shadorc commented:
I have reverted what you asked. I have left some imports that were removed because not used, do you want that I revert that too?
Also, I have seen that there was two identical enum ApplicationCommandOption.Type (core) and ApplicationCommandOptionType (REST), I have removed the first one.
